<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.340">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>process-models-workshop - 9&nbsp; Using MESS API mode in RStudio to explore hypotheses</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./ml_inference_classification.html" rel="next">
<link href="./mess_getting_started.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<script async="" src="https://hypothes.is/embed.js"></script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./mess_simulations.html"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Using MESS API mode in RStudio to explore hypotheses</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">process-models-workshop</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro_process_models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Theory and motivation of process modeling</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./coding_neutral_theory.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Coding exercise: A simple neutral model of biodiversity</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./neutral_theory_sims.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Simulation Exercise: Exploring neutral model behavior</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ml_inference_neutral.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Inferring the parameters of UNTB from outcomes</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./day_1_wrapup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Day 1 wrap-up</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./intro_mess.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Introduction to the MESS Model</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./mess_getting_started.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Getting started with the MESS model</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./mess_simulations.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Using MESS API mode in RStudio to explore hypotheses</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ml_inference_classification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">ML model selection with MESS simulations</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ml_inference_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">ML parameter estimation with MESS simulations</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./wrapup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Wrap-up</span></span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#key-questions" id="toc-key-questions" class="nav-link active" data-scroll-target="#key-questions"><span class="header-section-number">9.1</span> Key questions</a></li>
  <li><a href="#lesson-objectives" id="toc-lesson-objectives" class="nav-link" data-scroll-target="#lesson-objectives"><span class="header-section-number">9.2</span> Lesson objectives</a></li>
  <li><a href="#planned-exercises" id="toc-planned-exercises" class="nav-link" data-scroll-target="#planned-exercises"><span class="header-section-number">9.3</span> Planned exercises</a>
  <ul class="collapse">
  <li><a href="#import-mess-in-r-and-run-simulations-in-api-mode" id="toc-import-mess-in-r-and-run-simulations-in-api-mode" class="nav-link" data-scroll-target="#import-mess-in-r-and-run-simulations-in-api-mode"><span class="header-section-number">9.3.1</span> Import MESS in R and run simulations in API mode</a></li>
  <li><a href="#load-inspect-and-plot-simulation-results" id="toc-load-inspect-and-plot-simulation-results" class="nav-link" data-scroll-target="#load-inspect-and-plot-simulation-results"><span class="header-section-number">9.3.2</span> Load, inspect, and plot simulation results</a></li>
  <li><a href="#group-brainstorming-of-params-to-tweak" id="toc-group-brainstorming-of-params-to-tweak" class="nav-link" data-scroll-target="#group-brainstorming-of-params-to-tweak"><span class="header-section-number">9.3.3</span> Group brainstorming of params to tweak</a></li>
  <li><a href="#breakout-to-test-them" id="toc-breakout-to-test-them" class="nav-link" data-scroll-target="#breakout-to-test-them"><span class="header-section-number">9.3.4</span> Breakout to test them</a></li>
  <li><a href="#come-together-to-debriefsynthesize" id="toc-come-together-to-debriefsynthesize" class="nav-link" data-scroll-target="#come-together-to-debriefsynthesize"><span class="header-section-number">9.3.5</span> Come together to debrief/synthesize</a></li>
  </ul></li>
  <li><a href="#key-points" id="toc-key-points" class="nav-link" data-scroll-target="#key-points"><span class="header-section-number">9.4</span> Key points</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Using MESS API mode in RStudio to explore hypotheses</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="key-questions" class="level2" data-number="9.1">
<h2 data-number="9.1" class="anchored" data-anchor-id="key-questions"><span class="header-section-number">9.1</span> Key questions</h2>
<ol type="1">
<li>How do I run MESS from inside RStudio?</li>
<li>How do I get simulation results in a format that is useful?</li>
<li>How do I visualize simulation results to gain insight?</li>
<li>How do I use MESS to explore/generate theoretical predictions?</li>
</ol>
</section>
<section id="lesson-objectives" class="level2" data-number="9.2">
<h2 data-number="9.2" class="anchored" data-anchor-id="lesson-objectives"><span class="header-section-number">9.2</span> Lesson objectives</h2>
<p>After this lesson, learners should be able to…</p>
<ol type="1">
<li>Run MESS simulations in R</li>
<li>Summarize and plot simulation results</li>
<li>Formulate hypotheses for how tweaking parameters will affect model outputs</li>
<li>Use MESS to test these hypotheses in silico</li>
<li>Synthesize the outcomes</li>
</ol>
</section>
<section id="planned-exercises" class="level2" data-number="9.3">
<h2 data-number="9.3" class="anchored" data-anchor-id="planned-exercises"><span class="header-section-number">9.3</span> Planned exercises</h2>
<ul>
<li>Import MESS in R and run simulations in API mode</li>
<li>Load, inspect, and plot simulation results</li>
<li>Group brainstorming of params to tweak</li>
<li>Breakout to test them</li>
<li>Regroup to debrief/synthesize</li>
</ul>
<section id="import-mess-in-r-and-run-simulations-in-api-mode" class="level3" data-number="9.3.1">
<h3 data-number="9.3.1" class="anchored" data-anchor-id="import-mess-in-r-and-run-simulations-in-api-mode"><span class="header-section-number">9.3.1</span> Import MESS in R and run simulations in API mode</h3>
<p>The MESS CLI is a simple and powerful way to run simulations, but the <code>iBioGen</code> package also provies an API mode (in both R and python) which lets you dive under the hood of the CLI mode a bit. You have all the power of the CLI mode, yet more flexibility. In the source pane of RStudio open a new RScript and import the <code>iBioGen</code> package.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
The reticulate library
</div>
</div>
<div class="callout-body-container callout-body">
<p>A little bookkeeping is first required in loading the <code>reticulate</code> library, which is a binding layer between R and python, allowing seemless access to python modules from within R code.</p>
</div>
</div>
<div class="sourceCode" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reticulate)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="do">## Import the iBioGen package and give it an alias so we can refer to it as MESS</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>MESS <span class="ot">&lt;-</span> <span class="fu">import</span>(<span class="st">"iBioGen"</span>, <span class="at">as=</span><span class="st">"MESS"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The first step in API mode is to create a <code>Core</code> object. The <code>Core</code> object is the fundamental unit of the simulation, encompassing a standard Island Biogeography configuration with a very large <code>Metacommunity</code> and a much smaller <code>LocalCommunity</code> which is connected to it by colonization. In creating a <code>Core</code> object the only thing you’re required to pass in is a name, so lets go with “api-simdata”, to differentiate from the earlier <code>simdata</code> in CLI mode.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>core <span class="ot">=</span> MESS<span class="sc">$</span><span class="fu">Core</span>(<span class="st">"api-simdata"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>We can view the parameters of this new <code>core</code> object with the <code>get_params()</code> method:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="do">## verbose=TRUE will format the parameters in a nice way when printed</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>core<span class="sc">$</span><span class="fu">get_params</span>(<span class="at">verbose=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>------- iBioGen params file (v.0.0.9)-------------------------------------------</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>api-simdata          ## [0] [simulation_name]: The name of this simulation scenario</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>./default_iBioGen    ## [1] [project_dir]: Where to save files</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>1                    ## [2] [birth_rate]: Speciation rate</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>taxa                 ## [3] [meta_stop_criterion]: Whether to stop metacommunity on ntaxa or time</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>20                   ## [4] [ntaxa]: Number of taxa to simulate if stop is `ntaxa`</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>4                    ## [5] [time]: Amount of time to simulate if stop is `time`</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>abundance            ## [6] [process]: Whether to evolve `abundance` or growth `rate` via BM</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>True                 ## [7] [ClaDS]: Whether to allow speciation rate change a la ClaDS</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>50000                ## [8] [abundance_mean]: Ancestral abundance at time 0</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>0.1                  ## [9] [abundance_sigma]: Rate at which abundance changes if process is `abundance`</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>0                    ## [10] [growth_rate_mean]: Ancestral population growth rate at time 0.</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>0.01                 ## [11] [growth_rate_sigma]: Rate at which growth rate changes if process is `rate`</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>0.1                  ## [12] [ClaDS_sigma]: Rate at which speciation rate changes if ClaDS is True</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>0.9                  ## [13] [ClaDS_alpha]: Rate shift if ClaDS is True</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>500                  ## [14] [sequence_length]: Length of the genomic region simulated, in base pairs</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>1e-08                ## [15] [mutation_rate]: Mutation rate per base per generation</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>10                   ## [16] [sample_size]: Number of samples to draw for calculating genetic diversity</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>None                 ## [17] [abundance_scaling]: Scaling abundance to Ne (None, log, ln or a ratio)</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>1000                 ## [18] [J]: Number of individuals in the local community</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>0.005                ## [19] [colrate]: Rate of colonization into the local community (% / birth event)</span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a>neutral              ## [20] [assembly_model]: Selecting neutral or non-neutral assembly processes</span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a>2                    ## [21] [ecological_strength]: Impact of competition or filtering on fitness</span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>1                    ## [22] [generation_time]: Generation time of local community taxa (in years)</span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a>time                 ## [23] [local_stop_criterion]: Stop local community on time or equilibrium</span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a>100                  ## [24] [local_stop_time]: Local community simulation duration (in generations)</span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a>100                  ## [25] [tau_max]: Duration of anagenetic speciation (in generations)</span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a>10                   ## [26] [gene_flow_effect]: Damping effect of gene flow (in generations) on tau_max</span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a>1000                 ## [27] [Ne_scaling]: Scaling individuals to demes in local community</span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a>False                ## [28] [rm_duplicates]: Deduplicate seqs before calculating local community pi</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>These should all look somewhat familiar by now.</p>
<p>Changing parameters of <code>Core</code> objects can be accomplished with the <code>set_param()</code> method:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set a parameter to a fixed value</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>core<span class="sc">$</span><span class="fu">set_param</span>(<span class="st">"J"</span>, <span class="dv">5000</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Set a parameter to a prior range to be sampled uniformly</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>core<span class="sc">$</span><span class="fu">set_param</span>(<span class="st">"colrate"</span>, <span class="st">"0.01-0.05"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Finally, we can generate simulations by calling the <code>simulate()</code> method and specifying the number of simulations to perform:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Lets do 100 simulations so there's a _chance_ we'll see some pattern in</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="do">## in the data</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>core<span class="sc">$</span><span class="fu">simulate</span>(<span class="at">nsims=</span><span class="dv">100</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb7"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>    Generating 100.0 simulation(s).</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  [####################] 100%  Finished 100 simulations in   0:00:03| </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>If you are still in the <code>default_iBioGen</code> directory in your files pane, then you will see the new <code>api-simdata-SIMOUT.csv</code> file pop up there:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/MESS-Simulations-API-SSIMOUT.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">New api-simdata-SIMOUT.csv in the files pane</figcaption>
</figure>
</div>
<p>Now you might see a glimmer of how the API mode might allow you much more power in manipulating parameters and running simulations, but there is more!</p>
</section>
<section id="load-inspect-and-plot-simulation-results" class="level3" data-number="9.3.2">
<h3 data-number="9.3.2" class="anchored" data-anchor-id="load-inspect-and-plot-simulation-results"><span class="header-section-number">9.3.2</span> Load, inspect, and plot simulation results</h3>
<p>Recall that the output of MESS simulations is a file which contains parameters and data generated with one simulation per line. In the raw form, the simulated data are bundled up in a somewhat obscure format, with reasonable looking information in most columns, but obscure information in three columns: <code>metadata</code>, <code>metatree</code>, and <code>localdata</code>. The MESS API mode provides a method to unpack and <em>summarize</em> this obscure data into a way that is more meaningful and human readable (and also useful for plotting), and this is the <code>load_sims()</code> method.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load simulation results</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>results <span class="ot">=</span> core<span class="sc">$</span><span class="fu">load_sims</span>()</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Results is a 3 element list, each element is a data.frame that contains</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># one row per simulation</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="co"># results[[1]] - params and sumstats</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="co"># results[[2]] - Raw data per simulation (abunds, pis, traits, etc)</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="co"># results[[3]] - global phylogeny per simulation</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Lets focus for the moment on <code>results[[1]]</code> the data.frame of params and sumstats.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>results[[<span class="dv">1</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb10"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>   birth_rate meta_stop_criterion ntaxa time   process ClaDS abundance_mean abundance_sigma</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>1           1                taxa    20    4 abundance  TRUE          50000             0.1</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>2           1                taxa    20    4 abundance  TRUE          50000             0.1</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>3           1                taxa    20    4 abundance  TRUE          50000             0.1</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>4           1                taxa    20    4 abundance  TRUE          50000             0.1</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>5           1                taxa    20    4 abundance  TRUE          50000             0.1</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>6           1                taxa    20    4 abundance  TRUE          50000             0.1</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>7           1                taxa    20    4 abundance  TRUE          50000             0.1</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>8           1                taxa    20    4 abundance  TRUE          50000             0.1</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>9           1                taxa    20    4 abundance  TRUE          50000             0.1</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>10          1                taxa    20    4 abundance  TRUE          50000             0.1</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>   growth_rate_mean growth_rate_sigma ClaDS_sigma ClaDS_alpha sequence_length mutation_rate sample_size</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>1                 0              0.01         0.1         0.9             500         1e-08          10</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>2                 0              0.01         0.1         0.9             500         1e-08          10</span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>3                 0              0.01         0.1         0.9             500         1e-08          10</span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>4                 0              0.01         0.1         0.9             500         1e-08          10</span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>5                 0              0.01         0.1         0.9             500         1e-08          10</span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>6                 0              0.01         0.1         0.9             500         1e-08          10</span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>7                 0              0.01         0.1         0.9             500         1e-08          10</span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a>8                 0              0.01         0.1         0.9             500         1e-08          10</span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a>9                 0              0.01         0.1         0.9             500         1e-08          10</span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a>10                0              0.01         0.1         0.9             500         1e-08          10</span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a>   abundance_scaling    J    colrate assembly_model ecological_strength generation_time</span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a>1               None 5000 0.02816428        neutral                   2               1</span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a>2               None 5000 0.04468862        neutral                   2               1</span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a>3               None 5000 0.01764501        neutral                   2               1</span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a>4               None 5000 0.01780714        neutral                   2               1</span>
<span id="cb10-28"><a href="#cb10-28" aria-hidden="true" tabindex="-1"></a>5               None 5000 0.04374193        neutral                   2               1</span>
<span id="cb10-29"><a href="#cb10-29" aria-hidden="true" tabindex="-1"></a>6               None 5000 0.03645229        neutral                   2               1</span>
<span id="cb10-30"><a href="#cb10-30" aria-hidden="true" tabindex="-1"></a>7               None 5000 0.01470986        neutral                   2               1</span>
<span id="cb10-31"><a href="#cb10-31" aria-hidden="true" tabindex="-1"></a>8               None 5000 0.01802192        neutral                   2               1</span>
<span id="cb10-32"><a href="#cb10-32" aria-hidden="true" tabindex="-1"></a>9               None 5000 0.01327007        neutral                   2               1</span>
<span id="cb10-33"><a href="#cb10-33" aria-hidden="true" tabindex="-1"></a>10              None 5000 0.03985648        neutral                   2               1</span>
<span id="cb10-34"><a href="#cb10-34" aria-hidden="true" tabindex="-1"></a>   local_stop_criterion local_stop_time tau_max gene_flow_effect Ne_scaling rm_duplicates</span>
<span id="cb10-35"><a href="#cb10-35" aria-hidden="true" tabindex="-1"></a>1                  time             100     100               10       1000         FALSE</span>
<span id="cb10-36"><a href="#cb10-36" aria-hidden="true" tabindex="-1"></a>2                  time             100     100               10       1000         FALSE</span>
<span id="cb10-37"><a href="#cb10-37" aria-hidden="true" tabindex="-1"></a>3                  time             100     100               10       1000         FALSE</span>
<span id="cb10-38"><a href="#cb10-38" aria-hidden="true" tabindex="-1"></a>4                  time             100     100               10       1000         FALSE</span>
<span id="cb10-39"><a href="#cb10-39" aria-hidden="true" tabindex="-1"></a>5                  time             100     100               10       1000         FALSE</span>
<span id="cb10-40"><a href="#cb10-40" aria-hidden="true" tabindex="-1"></a>6                  time             100     100               10       1000         FALSE</span>
<span id="cb10-41"><a href="#cb10-41" aria-hidden="true" tabindex="-1"></a>7                  time             100     100               10       1000         FALSE</span>
<span id="cb10-42"><a href="#cb10-42" aria-hidden="true" tabindex="-1"></a>8                  time             100     100               10       1000         FALSE</span>
<span id="cb10-43"><a href="#cb10-43" aria-hidden="true" tabindex="-1"></a>9                  time             100     100               10       1000         FALSE</span>
<span id="cb10-44"><a href="#cb10-44" aria-hidden="true" tabindex="-1"></a>10                 time             100     100               10       1000         FALSE</span>
<span id="cb10-45"><a href="#cb10-45" aria-hidden="true" tabindex="-1"></a>   meta_obs_ntaxa meta_obs_time local_obs_time(gen) local_obs_time(eq) meta_turnover_rate   ClaDS_m</span>
<span id="cb10-46"><a href="#cb10-46" aria-hidden="true" tabindex="-1"></a>1              20      7.238911                 100             0.3724         0.20000000 0.9045113</span>
<span id="cb10-47"><a href="#cb10-47" aria-hidden="true" tabindex="-1"></a>2              20      3.304861                 100             0.6238         0.00000000 0.9045113</span>
<span id="cb10-48"><a href="#cb10-48" aria-hidden="true" tabindex="-1"></a>3              20      3.618550                 100             0.7324         0.04761905 0.9045113</span>
<span id="cb10-49"><a href="#cb10-49" aria-hidden="true" tabindex="-1"></a>4              20      2.464254                 100             0.4310         0.00000000 0.9045113</span>
<span id="cb10-50"><a href="#cb10-50" aria-hidden="true" tabindex="-1"></a>5              20      4.223441                 100             0.8360         0.00000000 0.9045113</span>
<span id="cb10-51"><a href="#cb10-51" aria-hidden="true" tabindex="-1"></a>6              20      2.421120                 100             0.8176         0.00000000 0.9045113</span>
<span id="cb10-52"><a href="#cb10-52" aria-hidden="true" tabindex="-1"></a>7              20      5.467468                 100             0.6822         0.00000000 0.9045113</span>
<span id="cb10-53"><a href="#cb10-53" aria-hidden="true" tabindex="-1"></a>8              20      2.940106                 100             0.4306         0.00000000 0.9045113</span>
<span id="cb10-54"><a href="#cb10-54" aria-hidden="true" tabindex="-1"></a>9              20      7.681525                 100             0.7238         0.00000000 0.9045113</span>
<span id="cb10-55"><a href="#cb10-55" aria-hidden="true" tabindex="-1"></a>10             20      5.754293                 100             0.9566         0.00000000 0.9045113</span>
<span id="cb10-56"><a href="#cb10-56" aria-hidden="true" tabindex="-1"></a>   local_S abund_h1 abund_h2 abund_h3    pi_h1    pi_h2    pi_h3  trait_h1 trait_h2 trait_h3</span>
<span id="cb10-57"><a href="#cb10-57" aria-hidden="true" tabindex="-1"></a>1       15 3.920161 2.333055 1.990338 5.600035 3.921122 3.358520  6.739850 4.842076 4.062254</span>
<span id="cb10-58"><a href="#cb10-58" aria-hidden="true" tabindex="-1"></a>2       16 7.128577 5.040473 4.096290 8.526984 7.054373 6.291502  7.802848 6.135959 5.476494</span>
<span id="cb10-59"><a href="#cb10-59" aria-hidden="true" tabindex="-1"></a>3       16 9.350507 7.267024 6.180439 7.649085 5.653503 4.763161 10.192978 8.298181 7.381126</span>
<span id="cb10-60"><a href="#cb10-60" aria-hidden="true" tabindex="-1"></a>4       20 4.587044 2.794328 2.309903 6.338057 5.330343 4.886343  6.694479 5.049368 4.660101</span>
<span id="cb10-61"><a href="#cb10-61" aria-hidden="true" tabindex="-1"></a>5       20 9.058059 6.625556 5.695691 7.261966 4.943116 4.148738 13.298536 9.102956 7.439101</span>
<span id="cb10-62"><a href="#cb10-62" aria-hidden="true" tabindex="-1"></a>6       18 4.943500 3.162048 2.632319 8.009983 6.529084 5.872421  8.282393 5.839957 4.894036</span>
<span id="cb10-63"><a href="#cb10-63" aria-hidden="true" tabindex="-1"></a>7       19 9.566252 6.654908 5.250049 9.211804 7.333543 6.414324 10.890139 8.424614 7.399630</span>
<span id="cb10-64"><a href="#cb10-64" aria-hidden="true" tabindex="-1"></a>8       15 4.710206 2.813355 2.309257 8.199976 6.782519 6.118322  9.273341 6.583071 5.696876</span>
<span id="cb10-65"><a href="#cb10-65" aria-hidden="true" tabindex="-1"></a>9       18 7.161965 5.070151 4.344310 5.869568 4.004179 3.236785 10.296570 8.314281 6.957238</span>
<span id="cb10-66"><a href="#cb10-66" aria-hidden="true" tabindex="-1"></a>10      17 6.138167 4.205529 3.658744 8.177872 6.203475 5.101742 11.851781 7.690572 5.713765</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Most of these things should look familiar (if not understandable), but there are several new columns which are very important. You might also notice at first that the <code>colrate</code> parameters per simulation do indeed vary, while all others are fixed, as we specified a prior range on this paramter.</p>
<p>The <strong>new</strong> thing to notice is that all the <strong>data</strong> (previously encapsulate in a human unreadable format) has been transformed into a much more user friendly summarization including <code>local_S</code> (indicating local species richness), and 9 columns of the form <code>*_h*</code>, which represent the first 3 Hill numbers on abundance (<code>abund</code>), nucleotide diversity (<code>pi</code>), and trait value distributions (<code>trait</code>).</p>
<p>Since we manipulated the <code>colrate</code> parameter in these simulations we can visualize the relationship between <code>colrate</code> and a few of the resulting summary statistics, like <code>local_S</code>, <code>abund_h1</code>, and <code>pi_h1</code> using the built-in R plotting function <code>plot()</code>.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Lets create a new variable for the results we're interested in</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>simes <span class="ot">&lt;-</span> results[[<span class="dv">1</span>]]</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Now plot local_S, abund_h1, and pi_h1 as a function of colrate</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sims<span class="sc">$</span>colrate, sims<span class="sc">$</span>local_S)</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sims<span class="sc">$</span>colrate, sims<span class="sc">$</span>abund_h1)</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sims<span class="sc">$</span>colrate, sims<span class="sc">$</span>pi_h1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><img src="images/MESS-Simulations-API-Jm20-localS.png" class="img-fluid" alt="Local species richness with ntaxa=20"> <img src="images/MESS-Simulations-API-Jm20-abund_h1.png" class="img-fluid" alt="First Hill number of local abundance with ntaxa=20"> <img src="images/MESS-Simulations-API-Jm20-pi_h1.png" class="img-fluid" alt="First Hill number of pi with ntaxa=20"></p>
<p>Here we can see a <em>weak</em> positive trend in all three summary statistics, with increasing <code>colrate</code> correlating with an increase in each of these data axes.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Discussion topic: Why would <code>colrate</code> impact these summary statistics in this way?
</div>
</div>
<div class="callout-body-container callout-body">
<p>Think for a moment about <strong>why</strong> increasing the colonization rate would increase local species richness, and the first Hill number of abundance and nucleotide diversity (on average). Let’s take a moment to reflect on these results.</p>
</div>
</div>
<p>It seems that local species richness is saturating, with uniformly high <code>S</code> which increasingly approaching 20 as <code>colrate</code> increases. What this means is that as <code>colrate</code> increases, the local community increasingly contains a total sample of the metacommunity, and there are no more unique lineages to colonize. Think back to what we learned yesterday about the neutral theory parameters. What parameter might we manipulate and how in order to relax this constraint and prevent the local community from saturating in this way?</p>
<p>::: {.callout-tip, collapse=“true”} ## How can we allow for more diversity in the local community?</p>
<p>Remember the <code>Jm</code> parameter from neutral theory, which increases the # of species in the metacommunity. In MESS we call this <code>ntaxa</code>, but it represents <em>exactly</em> the same parameter. What would happen if we increase the size of the metacommunity species pool (<code>ntaxa</code>) and also ran many more simulations?</p>
<p>Try increasing the value of <code>ntaxa</code> in your simulations, running it again, and plotting the results. How did this change affect <code>S</code>? :::</p>
<p>::: {.callout-tip, collapse=“true”} ## Solution</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>core<span class="sc">$</span><span class="fu">set_param</span>(<span class="st">"ntaxa"</span>, <span class="dv">100</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>core<span class="sc">$</span><span class="fu">simulate</span>(<span class="at">nsims=</span><span class="dv">500</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>results <span class="ot">=</span> core<span class="sc">$</span><span class="fu">load_sims</span>()</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>sims <span class="ot">&lt;-</span> results[[<span class="dv">1</span>]]</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sims<span class="sc">$</span>colrate, sims<span class="sc">$</span>local_S)</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sims<span class="sc">$</span>colrate, sims<span class="sc">$</span>abund_h1)</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sims<span class="sc">$</span>colrate, sims<span class="sc">$</span>pi_h1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>:::</p>
<p><img src="images/MESS-Simulations-API-Jm100-local_S.png" class="img-fluid" alt="Local species richness with ntaxa=100"> <img src="images/MESS-Simulations-API-Jm100-abund_h1.png" class="img-fluid" alt="First Hill number of local abundance with ntaxa=100"> <img src="images/MESS-Simulations-API-Jm100-pi_h1.png" class="img-fluid" alt="First Hill number of pi with ntaxa=100"></p>
</section>
<section id="group-brainstorming-of-params-to-tweak" class="level3" data-number="9.3.3">
<h3 data-number="9.3.3" class="anchored" data-anchor-id="group-brainstorming-of-params-to-tweak"><span class="header-section-number">9.3.3</span> Group brainstorming of params to tweak</h3>
</section>
<section id="breakout-to-test-them" class="level3" data-number="9.3.4">
<h3 data-number="9.3.4" class="anchored" data-anchor-id="breakout-to-test-them"><span class="header-section-number">9.3.4</span> Breakout to test them</h3>
</section>
<section id="come-together-to-debriefsynthesize" class="level3" data-number="9.3.5">
<h3 data-number="9.3.5" class="anchored" data-anchor-id="come-together-to-debriefsynthesize"><span class="header-section-number">9.3.5</span> Come together to debrief/synthesize</h3>
</section>
</section>
<section id="key-points" class="level2" data-number="9.4">
<h2 data-number="9.4" class="anchored" data-anchor-id="key-points"><span class="header-section-number">9.4</span> Key points</h2>
<ul>
<li>wat</li>
<li>do</li>
<li>chee</li>
</ul>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./mess_getting_started.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Getting started with the MESS model</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./ml_inference_classification.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">ML model selection with MESS simulations</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>